File: Notes/algorithims/sortincreasingfreq/quicksort.py
class Solution:

    def getFreq(self, arr):
        d = {}
        for num in arr:
            d[num] = d.get(num, 0) + 1
        return d
    
    def quickSort(self, arr, freq):
        if len(arr) <= 1:
            return arr
        
        pivot = arr[0]
        left = [x for x in arr[1:] if freq[x] < freq[pivot] or (freq[x] == freq[pivot] and x > pivot)]
        middle = [x for x in arr if x == pivot] # This is different than the org quick sort because we need to find dupes of pivot. In future problems for this just use this
        right = [x for x in arr[1:] if freq[x] > freq[pivot] or (freq[x] == freq[pivot] and x < pivot)]
        return self.quickSort(left, freq) + middle + self.quickSort(right, freq)
    

s = Solution()

l = [1, 3, 5, 7, 4, 2, 2, 4, 10]
d = s.getFreq(l)

print(s.quickSort(l, d))

File: Notes/DSA/algorithims/quicksort/quicksort.go
package main

import "fmt"

func quickSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}
	pivot := arr[len(arr)/2]
	var left, middle, right []int
	for _, x := range arr {
		switch {
		case x < pivot:
			left = append(left, x)
		case x == pivot:
			middle = append(middle, x)
		default:
			right = append(right, x)
		}
	}
	return append(append(quickSort(left), middle...), quickSort(right)...)
}

func main() {
	arr := []int{3, 6, 8, 10, 1, 2, 1}
	sortedArr := quickSort(arr)
	fmt.Println("Sorted array:", sortedArr)
}

//This implementation of quick sort has a time complexity of O(nlogn), on average.
// In the worst case, the time complexity is O(n^2), if the pivot is always the smallest or largest element in the array.

//The quick sort algorithm works by selecting a pivot element and partitioning the array into three subarrays: one containing
//elements less than the pivot, one containing elements equal to the pivot, and one containing elements greater than the pivot. It then recursively sorts the left and right subarrays.
